businessLogicLayer/Features/PricesNearMe/StationDataManager.swift

```diff
@@ -0,0 +1,123 @@
+//
+//  StationDataManager.swift
+//  fuelPrices
+//
+//  Created by Robson Harrison on 25/04/2024.
+//
+
+import Foundation
+
+// The StationDataManager is the API (or interface) to this feature of the app. The caller of this class should not need any other knowledge of any internal workings of HOW this class achieves its purpose.
+class StationDataManager: ObservableObject {
+    
+    enum StationDataManagerError: Error {
+        case stationDataAPIError(StationDataAPI.StationDataAPIError)
+        case refreshAlreadyInProgress
+    }
+    
+    let fuelProviders: [String: String]
+    private var dateOfLastRefresh: Date?
+    private(set) var isRefreshingData = false
+    // No layer above this class should know that StationDataAPI exists as a service.
+    private var stationDataAPI = StationDataAPI()
+    private var stationLibrary: [PetrolStation] = []
+    
+    // MARK: - Constructor
+    
+    init(fuelProviders: [String: String]) {
+        // a constructor (initialiser) should only initialize its properties. It shouldn't be used to trigger network calls or any other hidden behaviour.
+        self.fuelProviders = fuelProviders
+    }
+    
+    // MARK: - Publicly available functions / features (make it easy-to-use)
+    
+    func getStationData(forPostcode postcode: String, completion: @escaping (Result<[PetrolStation], StationDataManagerError>) -> Void) {
+        guard isRefreshingData == false else {
+            // Note: ALWAYS call your completion handlers in 100% of scenarios 😉
+            completion(.failure(.refreshAlreadyInProgress))
+            return
+        }
+        isRefreshingData = true
+        let endpointAddresses = fuelProviders.map{$1}
+        gatherStationData(from: endpointAddresses) { result in
+            DispatchQueue.main.async {
```

Comment:
We are ensuring that we are back on the main thread at this level (in the layered architecture of the app). This allows the networking calls below it to run on background threads and not have to care arbout UI-related restrictions. As this class is the interface for the feature we have decided to return values on the main thread.